DesignSafe-CI/portal
# task/DES-2641: Optionally render React components in the UI

## Overview: ##
Bootstraps an Nx project which can be used for replacing AngularJS pages with React. Adds working examples of hooks, and integration-style tests

## PR Status: ##


## Related Jira tickets: ##

* [DES-XXXX](https://tacc-main.atlassian.net/browse/DES-XXXX)

## Summary of Changes: ##
- Bootstraps an NX project, `client`, which will manage future frontend pieces written in React.
- Adds a hooks module (`@client/hooks`) for custom hooks. Includes an example hook for fetching the apps listing.
- Adds a component library for the Workspace (`@client/workspace`) with an example component that renders the apps listing.
- Adds support for testing using [Mock Service Worker](https://mswjs.io/) to simulate network calls during testing without the need to mock fetch/axios.
- Adds minimal routing using React-Router v6.
- Updates the Workspace template to include/render the Workspace section of the application.


## Testing Steps: ##
1. If you have a running local instance, take it down and run `docker-compose -f conf/docker/docker-compose-dev.all.debug.yml down`
2. Make sure you are using Node >= 20. (I use https://github.com/nvm-sh/nvm to manage this from the CLI). Since newer versions of node break our legacy angularJS build, run ```docker run -v `pwd`:`pwd` -w `pwd` -it node:16  /bin/bash -c "npm run dev"``` from the repo root to build the angularJS piece.
3. In your local `conf/env_files/designsafe.env` file, add the following line: 
`RENDER_REACT=True`
4. Open the `client` folder in a new editor window and run `npm ci` to install the new client dependencies..
5. With Designsafe running in local dev like usual, run `npm run start` from inside the `client` folder.
6. Visit https://designsafe.dev/rw/workspace/. You should see UI being rendered by Nx/React.
7. Run `npx nx run-many --all --target=test` from inside the `client` folder to confirm that the test harness works.

## UI Photos:
<img width="1189" alt="image" src="https://github.com/DesignSafe-CI/portal/assets/12601812/8ce59cc1-0442-4668-b091-de7e1be33f2a">


## Notes: ##
